README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Colorblind-friendly Palettes from Washington State

<!-- badges: start -->

[![R-CMD-check](https://github.com/CoryMcCartan/wacolors/workflows/R-CMD-check/badge.svg)](https://github.com/CoryMcCartan/wacolors/actions)
[![CRAN](https://www.r-pkg.org/badges/version/wacolors)](https://CRAN.R-project.org/package=wacolors)
<!-- badges: end -->

![Mt. Rainier](cover.jpg)

The `wacolor` package contains 28 color palettes taken from the
landscapes and cities of Washington state. Colors were extracted from a
set of photographs, and then combined to form a set of continuous and
discrete palettes. Continuous palettes were designed to be perceptually
uniform, while discrete palettes were chosen to maximize contrast at
several different levels of overall brightness and saturation. Each
palette has been evaluated to ensure colors are distinguishable by
colorblind people.

Discrete palettes contain at most seven colors. Don’t create graphics
that use more than seven discrete colors. You can color a map with four.
Anything more risks confusion. Consider differentiating through faceting
or labels, instead.

## Installation

You can install the released version of `wacolors` from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("wacolors")
```

You can install the development version of `wacolors` from GitHub with:

``` r
devtools::install_github("CoryMcCartan/wacolors")
```

## Usage

It’s easy to select a palette. Access palettes through `wacolors$...`
for autocompletion suggestions.

``` r
library(wacolors)
library(ggplot2)

# See all palettes
names(wacolors)
#>  [1] "rainier"         "washington_pass" "palouse"         "forest"         
#>  [5] "larch"           "coast"           "san_juan"        "uw"             
#>  [9] "fort_worden"     "skagit"          "flag"            "sound_sunset"   
#> [13] "ferries"         "forest_fire"     "sea"             "sea_star"       
#> [17] "volcano"         "baker"           "diablo"          "puget"          
#> [21] "mountains"       "gorge"           "foothills"       "footbridge"     
#> [25] "olympic"         "lopez"           "vantage"         "stuart"

# See one palette
wacolors$rainier
#>       lake    ragwort      lodge      trees     ground winter_sky paintbrush 
#>  "#465177"  "#E4C22B"  "#965127"  "#29483A"  "#759C44"  "#9FB6DA"  "#DF3383"
```

The package comes with `ggplot2`-compatible scales which are easy to
use.

``` r
# access by name
ggplot(mtcars, aes(mpg, wt)) +
    geom_point(aes(color = factor(cyl), size=hp)) +
    scale_color_wa_d("rainier")
```

![](man/figures/README-ggplot-1.png)<!-- -->

``` r
# or access using `wacolors$...`
ggplot(diamonds) +
    geom_bar(aes(x = cut, fill = clarity)) +
    scale_fill_wa_d(wacolors$sound_sunset, reverse=TRUE)
```

![](man/figures/README-ggplot-2.png)<!-- -->

You can use the `wa_pal` function to directly construct a
discretized/binned color scale.

``` r
image(volcano, col=wa_pal("ferries", 12))
```

![](man/figures/README-heatmap-1.png)<!-- -->

``` r
image(volcano, col=wa_pal("vantage", 200, reverse=TRUE))
```

![](man/figures/README-heatmap-2.png)<!-- -->

### Code Generation

In case you do not wish to have `wacolors` as a dependency, you may use
the `pal_vector()` and `pal_functions()` functions, which generate
self-contained code for using the palettes. When using RStudio, this
code will be loaded, ready to copy, at the console prompt.

``` r
pal_vector("sound_sunset", 15)
#> PAL_SOUND_SUNSET = c("#001E35", "#0E2648", "#342C59", "#523368", "#6D3D71",
#>                      "#804C73", "#925B78", "#A16C7B", "#BB797D", "#D08B79",
#>                      "#DE9F71", "#E6B465", "#EBCC5B", "#E6E55B", "#DCFF6C")

pal_functions("palouse")
#> scale_color_palouse_d = function(...) {
#>   pal_cols = c("#2D3F4A", "#C0A43D", "#8A6172", "#748A52", "#CCBA98",
#>                "#69A2E4")
#>   n_col = length(pal_cols)
#>   ramp = grDevices::colorRampPalette(pal_cols)
#>   pal_fun = function(n) if (n <= n_col) pal_cols[1:n] else ramp(n)
#>   discrete_scale("color", "palouse", palette=pal_fun, ...)
#> }
#> scale_fill_palouse_d = function(...) {
#>   pal_cols = c("#2D3F4A", "#C0A43D", "#8A6172", "#748A52", "#CCBA98",
#>                "#69A2E4")
#>   n_col = length(pal_cols)
#>   ramp = grDevices::colorRampPalette(pal_cols)
#>   pal_fun = function(n) if (n <= n_col) pal_cols[1:n] else ramp(n)
#>   discrete_scale("fill", "palouse", palette=pal_fun, ...)
#> }
#> 
#> scale_color_palouse_c = function(...) {
#>   scale_color_gradientn(..., colours=c("#2D3F4A", "#C0A43D", "#8A6172",
#>                         "#748A52", "#CCBA98", "#69A2E4"))
#> }
#> scale_fill_palouse_c = function(...) {
#>   scale_fill_gradientn(..., colours=c("#2D3F4A", "#C0A43D", "#8A6172",
#>                        "#748A52", "#CCBA98", "#69A2E4"))
#> }
#> 
#> 
```

## The Palettes

### Continuous Palettes

![](man/figures/README-cont-palettes-1.png)<!-- -->![](man/figures/README-cont-palettes-2.png)<!-- -->![](man/figures/README-cont-palettes-3.png)<!-- -->![](man/figures/README-cont-palettes-4.png)<!-- -->![](man/figures/README-cont-palettes-5.png)<!-- -->![](man/figures/README-cont-palettes-6.png)<!-- -->![](man/figures/README-cont-palettes-7.png)<!-- -->![](man/figures/README-cont-palettes-8.png)<!-- -->![](man/figures/README-cont-palettes-9.png)<!-- -->![](man/figures/README-cont-palettes-10.png)<!-- -->![](man/figures/README-cont-palettes-11.png)<!-- -->![](man/figures/README-cont-palettes-12.png)<!-- -->![](man/figures/README-cont-palettes-13.png)<!-- -->![](man/figures/README-cont-palettes-14.png)<!-- -->![](man/figures/README-cont-palettes-15.png)<!-- -->![](man/figures/README-cont-palettes-16.png)<!-- -->![](man/figures/README-cont-palettes-17.png)<!-- -->

### Discrete Palettes

![](man/figures/README-disc-palettes-1.png)<!-- -->![](man/figures/README-disc-palettes-2.png)<!-- -->![](man/figures/README-disc-palettes-3.png)<!-- -->![](man/figures/README-disc-palettes-4.png)<!-- -->![](man/figures/README-disc-palettes-5.png)<!-- -->![](man/figures/README-disc-palettes-6.png)<!-- -->![](man/figures/README-disc-palettes-7.png)<!-- -->![](man/figures/README-disc-palettes-8.png)<!-- -->![](man/figures/README-disc-palettes-9.png)<!-- -->![](man/figures/README-disc-palettes-10.png)<!-- -->![](man/figures/README-disc-palettes-11.png)<!-- -->


NEWS.md
# wacolors 0.3.0

* New diverging palettes: `lopez` and `stuart`
* New continuous palettes: `puget` and `footbridge`
* New discrete palette: `san_juan`
* Modified `foothills` and `baker` palettes
* Improve palette printing 
* Fix binned scales

# wacolors 0.2.1

* Improve documentation
* Slight modifications to existing palettes
* New `skagit` palette

# wacolors 0.2.0

* New, perceptually uniform palettes
* New discrete palettes systematically checked for CVD compatibility
* Support for changing midpoint of diverging palettes

# wacolors 0.1.3

* Fix tests

# wacolors 0.1.2

* Palette tweaks, new `forest_fire` palette.

# wacolors 0.1.0

* Initial release


tests/optim.R
if (F) {
# Tests and optimizations
library(colorspace)
library(ggplot2)


# sound_sunset ####
ss_old = c("#41527A", "#5C7293", "#7389A6", "#90929C",
           "#B09982", "#DDAD7A", "#F6E297")
sound_sunset = divergingx_hcl(15, h1=-122, h2=-6, h3=100, l1=9, l2 = 52, l3 = 95,
                              c1=18, cmax1=40, c2=33, cmax2=45, c3=88,
                              p1=1.9, p2=1, p3=0.8, p4=1)
specplot(sound_sunset, viridis::cividis(15))
specplot(sound_sunset, viridis::magma(15))
specplot(sound_sunset, viridis::inferno(15))
specplot(viridis::magma(15), viridis::inferno(15))
specplot(ss_old, sound_sunset[c(2,4,6,8,10,12,14)])
testpal(sound_sunset)
plot_hl(sound_sunset)
tvc(sound_sunset)

# ferries ####
f_old = c(deck="#2C303B", water="#3A4A6B", trim="#2C745F",
          windows="#8AA297", sun="#F4E8C8")
specplot(f_old)
ferries = divergingx_hcl(15, h1=275, h2=160, h3=60, l1=13, l2 = 54, l3 = 94,
                         c1=12, cmax1=50, c2=40, cmax2=32, c3=90,
                         p1=3, p2=1, p3=3.2, p4=1)
specplot(ferries, viridis::viridis(15))
specplot(ferries, colorRampPalette(f_old)(15))
plot_hl(ferries)
plot_hl(viridis::viridis(15))
tvc(ferries)


# sea_star ####
#sea_star = sequential_hcl(15, h=c(-50, 48), l=c(15, 80),
#                          c=c(35, 90), cmax=98, power=c(2.2, 1))
sea_star = divergingx_hcl(15, h1=-50, h2=25, h3=48, l1=15, l2 = 48, l3 = 81,
                          c1=35, cmax1=70, c2=35, cmax2=98, c3=90,
                          p1=0.8, p2=1, p3=0.6, p4=1)
specplot(sea_star, sound_sunset)
specplot(sea_star, viridis::rocket(15))
plot_hl(sea_star)
tvc(sea_star)


# forest_fire ####
ff_old = c("#3A2620", "#7A4424", "#CD5A0D", "#FE8B00", "#FFD874")
specplot(ff_old)
forest_fire = sequential_hcl(15, h=c(-5, 75), l=c(7, 93),
                            c=c(12, 76), cmax=120, power=c(1.1, 1))
specplot(forest_fire, viridis::inferno(15))
specplot(forest_fire, wa_pal("sea_star", 15))
plot_hl(forest_fire)
#specplot(forest_fire, ff_old)
tvc(forest_fire)

# sea ####
sea = sequential_hcl(15, h=c(270, 172), l=c(10, 89),
                     c=c(4, 22), cmax=52, power=c(2.9, 1))
specplot(sea, viridis::mako(15))
plot_hl(sea)
testpal(sea)
tvc(sea)


# volcano ####
v_old = c(rock="#30353F", ash="#9F9B93", glacier="#D9E1EC")
specplot(v_old)
volcano = divergingx_hcl(15, h1=280, h2=60, h3=300, l1=16, l2 = 53, l3 = 90,
                         c1=3, cmax1=0, c2=7, cmax2=0, c3=6,
                         p1=0.8, p2=1, p3=0.8, p4=1)
specplot(volcano, colorRampPalette(v_old)(15))
plot_hl(volcano)
testpal(volcano)
tvc(volcano)


# mountains ####
m_old = c("#161F47", "#134575", "#4A72B8", "#99C6F7", "#DCE8F8")
specplot(m_old)
mountains = sequential_hcl(15, h=c(210, 272), l=c(10, 80),
                           c=c(25, 32), cmax=65, power=c(1.1, 1))
specplot(mountains, colorRampPalette(m_old)(15))
testpal(mountains)
plot_hl(mountains)
plot_hl(sea)
tvc(mountains)

# baker ####
wm = c("#4D6E9F", "#9CB3DC", "#D0C5D1", "#F3ECE2", "#F1F1F1")
specplot(wm)
baker = sequential_hcl(15, h=c(-120, 80), l=c(52, 96),
                       c=c(30, 12), cmax=40, power=c(0.6, 1))
specplot(baker)
specplot(baker, colorRampPalette(wm)(15))
testpal(baker)
plot_hl(baker)
tvc(baker)


# diablo ####
diablo = sequential_hcl(15, h=c(120, 212), l=c(13, 77),
                        c=c(10, 37), cmax=52, power=c(3.2, 1))
specplot(diablo)
testpal(diablo)
plot_hl(diablo)
plot_hl(sea)
tvc(diablo)

# puget ####
puget = sequential_hcl(15, h=c(140, 375), l=c(18, 77),
                       c=c(10, 44), cmax=52, power=c(1.9, 1))
specplot(puget)
plot_hl(puget)
testpal(puget)
tvc(puget)

# gorge ####
gorge = sequential_hcl(15, h=c(170, 70), l=c(18, 82),
                        c=c(20, 30), cmax=56, power=c(1.15, 1))
specplot(deutan(gorge))
specplot(gorge)
testpal(gorge)
plot_hl(gorge)
tvc(gorge)

# foothills####
foothills = sequential_hcl(15, h=c(212, 90), l=c(5, 75),
                                  c=c(16, 8), cmax=68, power=c(0.7, 1.0))
specplot(foothills, gorge)
plot_hl(foothills)
testpal(foothills)
tvc(foothills)

# footbridge ####
footbridge = sequential_hcl(15, h=c(30, 65), l=c(15, 75),
                           c=c(10, 15), cmax=45, power=c(1.8, 1.0))
specplot(footbridge)
plot_hl(footbridge)
testpal(footbridge)
tvc(footbridge)

# olympic ####
olympic = divergingx_hcl(15, h1=120, h2=NA, h3=266, l1=18, l2=85, l3=18,
                         c1=10, cmax1=45, c2=5, cmax2=45, c3=15,
                         p1=1.4, p2=1, p3=1.4, p4=1) %>%
    rev()
specplot(olympic)
plot_hl(olympic)
testpal(olympic)
tvc(olympic)


# vantage ####
vantage = c(divergingx_hcl(8, h1=280, h2=270, h3=144,
                            l1=25, l2=60, l3=95,
                            c1=36, cmax1=46, c2=24, c3=10, p1=0.5, p3=0.8)[-8],
            hcl(96, 10, 95),
            divergingx_hcl(8, h1=84, h2=76, h3=24,
                           l1=95, l2=60, l3=25,
                           c1=32, cmax1=68, c2=52, cmax2=60, c3=44, p1=0.9, p3=0.4)[-1])
specplot(vantage)
testpal(vantage)
plot_hl(vantage)
tvc(vantage)


# lopez ####
lopez = diverging_hcl(15, h1=52, h2=265, l1=52, l2=95,
                      c=65, p1=0.4, p2=1)
specplot(lopez)
plot_hl(lopez)
testpal(lopez)
tvc(lopez)


# stuart ####
stuart = divergingx_hcl(15, h1=-80, h2=20, h3=140,
                        l1=27, l2=93, l3=27,
                        c1=45, cmax1=20, c2=25, cmax2=55, c3=15,
                        p1=0.7, p2=1.0, p3=0.8)
specplot(stuart)
plot_hl(stuart)
testpal(stuart)
tvc(stuart)


# HELPER FUNCTIONS --------------------------------------------------------

plot_hl = function(pal) {
    coords = as(hex2RGB(pal), "polarLUV")@coords
    ggplot(NULL, aes(x=coords[,"H"], y=coords[,"L"], color=pal)) +
        geom_point(size=3) +
        scale_color_identity() +
        coord_polar() +
        xlim(0, 360) +
        ylim(0, 100)
}

plot_ab = function(pal) {
    coords = as(hex2RGB(pal), "LAB")@coords
    ggplot(NULL, aes(x=coords[,"A"], y=coords[,"B"], color=pal)) +
        geom_point(size=3) +
        scale_color_identity()
}

tvc = function(palette) {
    varname = deparse(substitute(palette))
    code = paste0(varname, " = c(", paste0('"', palette, '"', collapse=", "), ")")
    code = paste0(strwrap(code, 76, indent=0, exdent=nchar(varname) + 5),
                  collapse="\n")
    code_output(code)
}

testpal = function(pal) {
    par(mfrow=c(2, 2), mar=rep(1, 4))
    hclplot(pal)
    demoplot(pal, "map")
    demoplot(pal, "scatter")
    demoplot(pal, "spine")
}

test_all = function(type="map") {
    dim = ceiling(sqrt(length(wacolors)))
    par(mfrow=c(dim, dim), mar=rep(0.2, 4))
    for (pal in wacolors) {
        demoplot(pal, type)
    }
}

}


tests/testthat.R
library(testthat)
library(wacolors)

test_check("wacolors")


tests/testthat/test_wacolors.R
# Very basic tests

test_that("Plotting works OK", {
    x = c("red", "black")
    attr(x, "name") = "pal"

    expect_invisible(plot.palette(x))
})

test_that("Code generation works OK", {
    exp_out = 'PAL_RAINIER = c('
    expect_output(pal_vector("rainier", 4), exp_out, fixed=TRUE)

})


R/util.R
#' Takes name of palette or palette itself and returns palette + name
#' @param name the name of the palette or the palette itself
#'
#' @keywords internal
#' @noRd
match_pal = function(name) {
    found_name = pmatch(name, names(wacolors))
    if (length(found_name) > 0 && !any(is.na(found_name))) {
        pal = wacolors[[found_name]]
    } else if (all(is.na(found_name)) && is.character(name) &&
               all(startsWith(name, "#"))) {
        pal = name
        try_match = vapply(wacolors, FUN=function(x) isTRUE(all.equal(x, pal)),
                           FUN.VALUE=logical(1))
        name = names(wacolors)[try_match]
    } else {
        stop("Palette `", name, "` not found.")
    }

    list(pal=pal, name=name)
}

#' @export
#' @importFrom graphics rect par image text
print.palette = function(x, ...) {
    x_print = unclass(x)
    if (!is.null(pal_name <- attr(x, "name", exact=TRUE))) {
        cat("`", pal_name, "`\n", sep="")
        attr(x_print, "name") = NULL
    }
    print(x_print)

    if (requireNamespace("cli", quietly=TRUE)) {
        n_colors = cli::num_ansi_colors()
        if (n_colors >= 256) {
            if (n_colors < 256^3/2) cat("approximate palette:\n")
            block = "\u00A0\u00A0"
            cat(" ")
            for (color in x) {
                cat(cli::make_ansi_style(color, bg=TRUE)(block))
                cat(" ")
            }
        }
    }

    cat("\n")

    invisible(x)
}

# `plot.palette` modified from that in `wesanderson` (c) 2016 Karthik Ram
#' @export
#' @importFrom graphics rect par image text
plot.palette = function(x, ...) {
    n <- length(x)
    old <- par(mar=c(0.5, 0.5, 0.5, 0.5))
    on.exit(par(old))

    image(1:n, 1, as.matrix(1:n), col=x,
          ylab="", xaxt="n", yaxt="n", bty="n")

    rect(0, 0.9, n + 1, 1.1, col=grDevices::rgb(1, 1, 1, 0.8), border=NA)
    text((n + 1) / 2, 1, labels=attr(x, "name", exact=TRUE), col="black", cex=1, font=2)

    if (!is.null(names(x)))
        text(1:n, 1.25, labels=names(x), col="black", cex=1)
}

#' Paste-to-hex
#' @keywords internal
#' @noRd
pth = function(x) {
    rgb = as.list(as.integer(strsplit(x, "\\s")[[1]]) / 256)
    col = do.call(colorspace::sRGB, rgb)
    str = paste0('"', colorspace::hex(col), '"')
    code_output(str)
}

# Helper for interactive code output
code_output = function(x) {
    if (interactive() && requireNamespace("rstudioapi", quietly=TRUE)) {
        rstudioapi::sendToConsole(x, execute=F)
    } else {
        cat(x, "\n")
    }
}


R/scales.R
#' Helper function-palette for discrete scales
#'
#' @param pal the palette colors
#' @param reverse whether to reverse
#'
#' @importFrom grDevices colorRampPalette
#'
#' @keywords internal
#' @noRd
discr_pal = function(pal, reverse=FALSE) {
    n_col = length(pal)
    names(pal) = NULL
    ramp = colorRampPalette(pal, space="Lab", interpolate="spline")
    function(n) {
        if (n <= n_col)
            pal[1:n]
        else
            ramp(n)
    }
}

#' Color palettes for `ggplot2`
#'
#' @rdname scale_wa
#'
#' @param palette a [`wacolors`] palette or palette name.
#' @param which if not `NULL`, the indices or names of a subset of colors to use.
#' @param midpoint if not `NULL` and at least one limit is not provided, the
#'   value to center the scale at. Useful for diverging scales.
#' @param ... Other arguments passed on to [ggplot2::discrete_scale()],
#'   [ggplot2::continuous_scale()], or [ggplot2::binned_scale()] to control
#'   name, limits, breaks, labels and so forth.
#' @param reverse `TRUE` if the colors should be reversed.
#'
#' @return A [ggplot2::Scale] object.
#'
#' @examples
#' library(ggplot2)
#'
#' ggplot(mtcars, aes(mpg, wt)) +
#'     geom_point(aes(color = factor(cyl), size=hp)) +
#'     scale_color_wa_d()
#'
#' ggplot(mtcars, aes(mpg, wt)) +
#'     geom_point(aes(color = hp)) +
#'     scale_color_wa_c("palouse",  which=c("snake", "wheat"))
#'
#' ggplot(diamonds) +
#'     geom_bar(aes(x = cut, fill = clarity)) +
#'     scale_fill_wa_d(wacolors$sound_sunset, reverse=TRUE)
#'
#' @importFrom ggplot2 discrete_scale binned_scale scale_color_gradientn scale_fill_gradientn
#' @export
scale_color_wa_d = function(palette="rainier", which=NULL, ..., reverse=FALSE) {
    pal = match_pal(palette)
    if (!is.null(which)) pal$pal = pal$pal[which]
    if (reverse) pal$pal = rev(pal$pal)

    if (pal$name %in% cont_pal)
        pal_fun = colorRampPalette(pal$pal, space="Lab", interpolate="spline")
    else
        pal_fun = discr_pal(pal$pal)

    discrete_scale("colour", pal$name, palette=pal_fun, ...)
}

#' @rdname scale_wa
#' @export
scale_fill_wa_d = function(palette="rainier", which=NULL, ..., reverse=FALSE) {
    pal = match_pal(palette)
    if (!is.null(which)) pal$pal = pal$pal[which]
    if (reverse) pal$pal = rev(pal$pal)

    if (pal$name %in% cont_pal)
        pal_fun = colorRampPalette(pal$pal, space="Lab", interpolate="spline")
    else
        pal_fun = discr_pal(pal$pal)

    discrete_scale("fill", pal$name, palette=pal_fun, ...)
}

#' @rdname scale_wa
#' @export
scale_color_wa_c = function(palette="sound_sunset", which=NULL, midpoint=NULL,
                            ..., reverse=FALSE) {
    pal = match_pal(palette)
    if (!is.null(which)) pal$pal = pal$pal[which]
    if (reverse) pal$pal = rev(pal$pal)

    if (is.null(midpoint)) {
        rescaler = scales::rescale
    } else {
        rescaler = function(x, to=c(0, 1), from=range(x, na.rm=TRUE)) {
            scales::rescale_mid(x, to, from, midpoint)
        }
    }
    scale_color_gradientn(..., colours=pal$pal, rescaler=rescaler)
}

#' @rdname scale_wa
#' @export
scale_fill_wa_c = function(palette="sound_sunset", which=NULL, midpoint=NULL,
                           ..., reverse=FALSE) {
    pal = match_pal(palette)
    if (!is.null(which)) pal$pal = pal$pal[which]
    if (reverse) pal$pal = rev(pal$pal)

    if (is.null(midpoint)) {
        rescaler = scales::rescale
    } else {
        rescaler = function(x, to=c(0, 1), from=range(x, na.rm=TRUE)) {
            scales::rescale_mid(x, to, from, midpoint)
        }
    }
    scale_fill_gradientn(..., colours=pal$pal, rescaler=rescaler)
}

#' @rdname scale_wa
#' @export
scale_color_wa_b = function(palette="sound_sunset", which=NULL, ..., reverse=FALSE) {
    pal = match_pal(palette)
    if (!is.null(which)) pal$pal = pal$pal[which]
    if (reverse) pal$pal = rev(pal$pal)
    pal_fun = scales::colour_ramp(pal$pal)

    binned_scale("color", pal$name, palette=pal_fun, ...)
}

#' @rdname scale_wa
#' @export
scale_fill_wa_b = function(palette="sound_sunset", which=NULL, ..., reverse=FALSE) {
    pal = match_pal(palette)
    if (!is.null(which)) pal$pal = pal$pal[which]
    if (reverse) pal$pal = rev(pal$pal)
    pal_fun = scales::colour_ramp(pal$pal)

    binned_scale("fill", pal$name, palette=pal_fun, ...)
}

#' @rdname scale_wa
#' @export
scale_colour_wa_d = scale_color_wa_d
#' @rdname scale_wa
#' @export
scale_colour_wa_c = scale_color_wa_c
#' @rdname scale_wa
#' @export
scale_colour_wa_b = scale_color_wa_b


R/colors.R
#' Washington State Color Palettes
#'
#' A collection of colorblind-friendly color palettes for various settings in
#' the state of Washington. Colors were extracted from a set of photographs, and
#' then combined to form a set of continuous and discrete palettes.  Continuous
#' palettes were designed to be perceptually uniform, while discrete palettes
#' were chosen to maximize contrast at several different levels of overall
#' brightness and saturation. Each palette has been evaluated to ensure colors
#' are distinguishable by colorblind people.
#'
#' Discrete palettes contain at most seven colors.  Don't create graphics that
#' use more than seven discrete colors.  You can color a map with four.
#' Anything more risks confusion.  Consider differentiating through faceting or
#' labels, instead.
#'
#' **Available continuous palettes**:
#'
#' `r paste0('![](README-cont-palettes-', seq_along(cont_pal), '.png)\n', collapse="\n")`
#'
#' **Available discrete palettes**:
#'
#' `r paste0('![](README-disc-palettes-', seq_len(length(wacolors) - length(cont_pal)), '.png)\n', collapse="\n")`
#'
#' @format A list of character vectors containing the color palettes. Discrete
#' palette vectors contain names for each color.
#'
#' @examples
#' wacolors$rainier
#' wacolors$palouse[1:4]
#'
#' @export
wacolors = list(
    # DISCRETE
    rainier = c(lake="#465177", ragwort="#E4C22B", lodge="#965127",
               trees="#29483A", ground="#759C44", winter_sky="#9FB6DA",
               paintbrush="#DF3383"),
    washington_pass = c(trees="#31543B", stone="#48628A", tips="#94AA3D",
                        road="#7F9CB1", sunbreak="#D9D1BE", stump="#3E3C3A"),
    palouse = c(snake="#2D3F4A", wheat="#C0A43D", fallow="#8A6172",
                hills="#748A52", canyon="#CCBA98", sky="#69A2E4"),
    forest = c(trees="#254029", stream="#1E212F", fern="#516F25",
               bark="#3A270A", mountains="#40666F"),
    larch = c(larch="#D2A554", shrub="#626B5D", rock="#8C8F9E",
              moss="#858753", sky="#A4BADF", dirt="#D3BEAF"),
    coast = c(surf="#7BAEA0", sea="#386276", rocks="#3A4332", sand="#7A7D6F",
              sunset="#D9B96E", sky="#BED4F0"),
    san_juan = c(trees="#21281D", grass="#CA884C", sea="#3A5775",
                 driftwood="#BAAF9F", clouds="#C9DCE2"),
    uw = c(purple="#483778", gold="#D2C28B", brick="#7E4837", cherry="#D48792",
           stone="#6B7471"),
    fort_worden = c(sea="#4D5370", shrub="#263C19", lighthouse="#A6B7C6",
                    rust="#8F6D3F"),
    skagit = c(red="#B51829", yellow="#EFC519", violet="#831285",
               orange="#CC7810", purple="#C886E8", mountains="#3E6E94"),
    flag = c(green="#247F5B", yellow="#E2CD70", blue="#5DB3DB", tan="#FFDCCD"),
    # CONTINUOUS
    sound_sunset = c("#001E36", "#0F2649", "#352C5A", "#533369", "#6E3D71",
                     "#814C74", "#925C78", "#A26D7C", "#BC7A7D", "#D08B79",
                     "#DE9F71", "#E7B565", "#EBCC5C", "#E7E55C", "#DCFF6C"),
    ferries = c("#241E33", "#1E2C54", "#003F69", "#005373", "#006372",
                "#007273", "#008174", "#349075", "#579E78", "#77AB7E",
                "#97B788", "#B3C493", "#CFD28F", "#F2DE83", "#FFE66C"),
    forest_fire = c("#2C0915", "#40111D", "#551A23", "#6A2626", "#7E3225",
                    "#92401D", "#A64F00", "#B96000", "#CB7100", "#DB8300",
                    "#E69800", "#EDAD00", "#F3C307", "#FAD753", "#FFEC7A"),
    sea = c("#1C1B23", "#222541", "#1F325A", "#094270", "#005282", "#0B6184",
            "#2F6F89", "#457D90", "#598C99", "#6B9AA3", "#7CA9AE", "#8DB9B9",
            "#9DC8C5", "#ADD8D2", "#BEE8DF"),
    sea_star = c("#4B0049", "#5D014F", "#700853", "#821554", "#932252",
                 "#963D4E", "#985350", "#95675B", "#B16D51", "#C3774D",
                 "#D48349", "#E48F43", "#F29C3B", "#FFAA3B", "#FFBA54"),
    volcano = c("#29272C", "#323335", "#3E3F3F", "#4A4B4B", "#555857",
                "#626561", "#71726B", "#837E77", "#8C8D87", "#989B98",
                "#A7A8A8", "#B5B7B7", "#C2C5C8", "#D3D3D9", "#E7E0E8"),
    baker = c("#627F9A", "#7684A7", "#8B88B1", "#A08CB9", "#B490BF", "#C795C3",
              "#D69CC3", "#E1A5C0", "#E9AFBE", "#EEBABD", "#F2C6BE", "#F4D2C2",
              "#F4DEC9", "#F5E9D3", "#F6F4E5"),
    diablo = c("#172512", "#0E320C", "#003F0F", "#004C1A", "#005929",
               "#00663C", "#007051", "#007B63", "#008575", "#279085",
               "#3E9B96", "#51A6A6", "#63B2B7", "#75BDC7", "#87C8D8"),
    puget = c("#1D3024", "#123C2E", "#00483E", "#005352", "#005C67", "#01657D",
              "#386B91", "#5F6FA3", "#8272B1", "#A175B8", "#B87DB8", "#CB86B6",
              "#DA92B3", "#E69FAF", "#EFAEAB"),
    mountains = c("#002733", "#003141", "#003B50", "#004661", "#005172",
                  "#005C84", "#006797", "#0071AA", "#3A7BB5", "#5E85BE",
                  "#7990C7", "#8F9CCF", "#A4A8D8", "#B7B4E1", "#C8C2EA"),
    gorge = c("#00352A", "#00402F", "#004C33", "#005736", "#076338", "#236E38",
              "#387937", "#4D8434", "#658E3C", "#7C984D", "#91A15E", "#A5AB70",
              "#B7B583", "#C8C095", "#D9CBA7"),
    foothills = c("#001C28", "#00262D", "#003233", "#003E39", "#004B3E",
                  "#005841", "#006542", "#00723F", "#007F38", "#18893C",
                  "#529249", "#739C5B", "#8FA572", "#A6AF8C", "#B9BAAF"),
    footbridge = c("#32221C", "#42291C", "#50311D", "#5E391D", "#6B431C",
                   "#774C1E", "#7D5933", "#846545", "#8B7155", "#927C64",
                   "#9A8873", "#A39481", "#ACA08E", "#B6AC9B", "#C1B8A7"),
    olympic = c("#2A2A40", "#3E3E63", "#535285", "#6B6B98", "#8484A4",
                "#9E9EB4", "#B9B9C5", "#D4D4DA", "#B3BDB0", "#93A58E",
                "#738F6A", "#517942", "#396128", "#2D4823", "#232F1F"),
    lopez = c("#A3730F", "#B2833A", "#C19456", "#D0A570", "#DEB689", "#EBC8A3",
              "#F6DBBF", "#F1F1F1", "#DBDCFF", "#C9CAFA", "#B7B8F0", "#A5A7E5",
              "#9496DA", "#8385CE", "#7275C2"),
    vantage = c("#473065", "#5B497F", "#70629B", "#857EAB", "#949EB5",
                "#A4BDC6", "#C0DAD6", "#EFF2E5", "#DCD797", "#C5BB61",
                "#AD9F3C", "#9A8224", "#89641E", "#774718", "#632B14"),
    stuart = c("#4F3173", "#6F457F", "#8D5C8C", "#A87499", "#C18FA8",
               "#D8AAB8", "#EDC7CB", "#FFE2DE", "#F6C7AE", "#D8B07A",
               "#AF9C57", "#85884B", "#617344", "#425C3D", "#2C4534")
)

cont_pal = c("sound_sunset", "ferries", "forest_fire", "sea_star", "sea",
             "volcano", "baker", "diablo", "puget", "mountains", "gorge",
             "foothills", "footbridge", "olympic", "lopez", "vantage", "stuart")


#' Washington State Color Palette Generator
#'
#' Generate `palette` objects from the `wacolors` list
#'
#' @param palette The name of the palette (partial matching supported), or an
#'   actual palette from `[wacolors]`.
#' @param which if not `NULL`, the indices or names of a subset of colors to use.
#' @param n The number of colors in the palette. If this exceeds the actual
#'   number and `type` is not provided, it will be set to `continuous`.
#' @param type Either `continuous` or `discrete`. Use `continuous` if you want
#'   to automatically interpolate between colors.
#' @param reverse `TRUE` if palette should be reversed.
#'
#' @return A vector of colors of type `palette`. Use the `plot()` function to
#'   plot the palette. If the `cli` package is installed, printing the palette
#'   to the console will also show its colors.
#'
#' @examples
#' wa_pal("rainier")
#' wa_pal(wacolors$rainier)
#' wa_pal("sound_sunset", 20, "continuous")
#' wa_pal("washington_pass", reverse=TRUE)
#'
#' @importFrom grDevices colorRampPalette
#' @export
wa_pal = function(palette, n, which=NULL,
                  type=c("discrete", "continuous"), reverse=FALSE) {
    obj = match_pal(palette)
    pal = obj$pal
    name = obj$name
    if (!is.null(which)) pal = pal[which]

    if (is.null(pal))
        stop("Palette `", name, "` not found.")

    if (missing(n))
        n = length(pal)

    if (name %in% cont_pal || n > length(pal))
        type="continuous"
    type = match.arg(type)

    if (type == "discrete")
        out = pal[1:n]
    else
        out = colorRampPalette(pal, space="Lab", interpolate="spline")(n)

    if (reverse) out = rev(out)

    structure(out, class="palette", name=name)
}
